eclair-core/src/test/scala/fr/acinq/eclair/io/PeerReadyNotifierSpec.scala

```diff
@@ -0,0 +1,172 @@
+    import f._
+
+    val notifier = testKit.spawn(PeerReadyNotifier(remoteNodeId, switchboard.ref, timeout_opt = Some(Right(BlockHeight(500)))))
+    notifier ! NotifyWhenPeerReady(probe.ref)
+    val request = switchboard.expectMessageType[Switchboard.GetPeerInfo]
+    request.replyTo ! Peer.PeerInfo(TestProbe().ref.toClassic, remoteNodeId, Peer.CONNECTED, None, Set.empty)
+    probe.expectMessage(PeerReady(remoteNodeId, 0))
+  }
+
+  test("peer connected (with channels)") { f =>
+    import f._
+
+    val notifier = testKit.spawn(PeerReadyNotifier(remoteNodeId, switchboard.ref, timeout_opt = Some(Right(BlockHeight(500)))))
+    notifier ! NotifyWhenPeerReady(probe.ref)
+    val request = switchboard.expectMessageType[Switchboard.GetPeerInfo]
+    request.replyTo ! Peer.PeerInfo(TestProbe().ref.toClassic, remoteNodeId, Peer.CONNECTED, None, channels)
+
+    // Channels are not ready yet.
+    channelProbes.foreach(_.expectMessageType[CMD_GET_CHANNEL_STATE].replyTo ! RES_GET_CHANNEL_STATE(SYNCING))
+    probe.expectNoMessage(100 millis)
+
+    // After the first retry, one of the channels is ready but not the second one.
+    channelProbes.head.expectMessageType[CMD_GET_CHANNEL_STATE].replyTo ! RES_GET_CHANNEL_STATE(NORMAL)
+    channelProbes.last.expectMessageType[CMD_GET_CHANNEL_STATE].replyTo ! RES_GET_CHANNEL_STATE(SYNCING)
+    probe.expectNoMessage(100 millis)
+
+    // After the second retry, both channels are ready.
+    channelProbes.head.expectMessageType[CMD_GET_CHANNEL_STATE].replyTo ! RES_GET_CHANNEL_STATE(NORMAL)
+    channelProbes.last.expectMessageType[CMD_GET_CHANNEL_STATE].replyTo ! RES_GET_CHANNEL_STATE(SHUTDOWN)
+    probe.expectMessage(PeerReady(remoteNodeId, 2))
+  }
+
+  test("peer connects after initial request") { f =>
+    import f._
+
+    val notifier = testKit.spawn(PeerReadyNotifier(remoteNodeId, switchboard.ref, timeout_opt = Some(Right(BlockHeight(500)))))
+    notifier ! NotifyWhenPeerReady(probe.ref)
+    val request1 = switchboard.expectMessageType[Switchboard.GetPeerInfo]
+    request1.replyTo ! Peer.PeerInfo(TestProbe().ref.toClassic, remoteNodeId, Peer.DISCONNECTED, None, channels)
+    channelProbes.head.expectNoMessage(100 millis)
+
+    // An unrelated peer connects.
+    system.eventStream ! EventStream.Publish(PeerConnected(TestProbe().ref.toClassic, randomKey().publicKey, null))
+    switchboard.expectNoMessage(100 millis)
+
+    // The target peer connects.
+    system.eventStream ! EventStream.Publish(PeerConnected(TestProbe().ref.toClassic, remoteNodeId, null))
+    val request2 = switchboard.expectMessageType[Switchboard.GetPeerInfo]
+    request2.replyTo ! Peer.PeerInfo(TestProbe().ref.toClassic, remoteNodeId, Peer.CONNECTED, None, channels)
+    channelProbes.foreach(_.expectMessageType[CMD_GET_CHANNEL_STATE].replyTo ! RES_GET_CHANNEL_STATE(NEGOTIATING))
+    probe.expectMessage(PeerReady(remoteNodeId, 2))
+  }
+
+  test("peer connects then disconnects") { f =>
```

Comment:
Related to my other comment about `ChannelStatesCollector` - could this situation lead to many actors being spawned from the  `waitForChannelsReady` state?

Reply:
This is a good point, fixed in https://github.com/ACINQ/eclair/pull/2464/commits/91f13eddd578bb68f3ac9614d8da2c66d1b9a51b

Reply:
Change/test solved any concerns.